File: bitwiseOOS.cpp
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    int subarrayBitwiseORs(vector<int>& arr) {
        unordered_set<int> result;
        unordered_set<int> curr;

        for (int num : arr) {
            unordered_set<int> next;
            next.insert(num);
            for (int val : curr)
                next.insert(val | num);
            curr = std::move(next);
            result.insert(curr.begin(), curr.end());
        }

        return result.size();
    }
};

File: bitwiseOOS.rs
use std::collections::HashSet;

impl Solution {
    pub fn subarray_bitwise_o_rs(arr: Vec<i32>) -> i32 {
        let mut result: HashSet<i32> = HashSet::new();
        let mut curr: HashSet<i32> = HashSet::new();

        for &num in &arr {
            let mut next: HashSet<i32> = HashSet::new();
            next.insert(num);

            for &val in &curr {
                next.insert(val | num);
            }

            curr = next;
            result.extend(&curr);
        }

        result.len() as i32
    }
}
